numpy/core/numeric.py

```diff
@@ -18,7 +18,7 @@
     _fastCopyAndTranspose as fastCopyAndTranspose, ALLOW_THREADS,
     BUFSIZE, CLIP, MAXDIMS, MAY_SHARE_BOUNDS, MAY_SHARE_EXACT, RAISE,
     WRAP, arange, array, broadcast, can_cast, compare_chararrays,
-    concatenate, copyto, count_nonzero, dot, dtype, empty,
```

Comment:
I'm not sure we can remove this.

I started trying to go down the road of trying to clear up inports, and realized that we have a problem - we'll break anyone using `np.core.numeric.count_nonzero` in their code. It seems to me we're stuck importing piles of junk...

The alternative would be to declare _"if it's not in `__all__`, it's subject to move at any time.

Reply:
+1 for removing things not in `__all__`. It is so much simpler to spell `np.core.numeric.count_nonzero` as `np.count_nonzero`. Does this need a release note? We should at least have a clear namespace policy - do we continue to hold onto `np.core.numeric`, `np.core.multiarray`, `np.core.umath` or do we export names directly from `np`?

Reply:
Perhaps this should hit the mailing list.

Perhaps the easiest option is just to:
* move `np.core.numeric` to `np.core._numeric`, and similar for other modules within it. The underscore would make the "don't rely on this" intent
* create a new `np.core.numeric` that:
  * emits a warning upon import
  * preserves the same api as we currently have

This would give us the freedom to come up with a more logical file structure, without risk of breaking anything

I'd still advocate leaving `np.core` around as a place for semi-public internals for other infrastructure packages (ie `np.core.normalize_axis_index`)

Reply:
IIRC, `ma.testutils` suffers from that problem, and lots of people were importing function from there that they shouldn't have.

Reply:
+1 for Eric's [proposal](https://github.com/numpy/numpy/pull/11909#discussion_r216148179)

> I think such a proposal would probably be NEP-worthy

agree

> The alternative would be to declare "if it's not in `__all__`, it's subject to move at any time."

In principle I'd say:
- the bar is a lot lower there
- we shouldn't (re)move things gratuitously
- it's okay to do so though if it would take a lot of work to keep non-public API in place or if it would limit evolution of the code base.

Reply:
NEP draft at #11924